floodscan_validation/exploration/threshold_deciding.R

```diff
@@ -0,0 +1,1083 @@
+################################
+
+# get station activations and basins together
+basin_activation <- sf_gauges_filtered %>%
+  st_join(
+    sf_basins %>%
+      filter(
+        HYBAS_LEVEL == 4
+      ),
+    join = st_intersects
+  ) %>%
+  select(
+    gauge_id, basin_id = HYBAS_ID
+  ) %>%
+  inner_join(
+    google_gauge_activations,
+    by = "gauge_id"
+  ) %>%
+  as_tibble() %>%
+  select(
+    -geometry
+  )
+
+# simple approach, take max avlue of each year we have full data for
+# and then take the median, which will ensure half the years would trigger
+# and half would not for an approx 2 year RP thresholding
+fs_basin_thresholds <- fs_basin %>%
+  mutate(
+    year = lubridate::year(time)
+  ) %>%
+  filter(
+    year %in% 1999:2022
+  ) %>%
+  group_by(
+    HYBAS_ID,
+    year
+  ) %>%
+  summarize(
+    sfed_max = max(sfed),
+    .groups = "drop_last"
+  ) %>%
+  summarize(
+    sfed_threshold = median(sfed_max),
+    .groups = "drop"
+  )
+
+# now check when the basins cross these thresholds
+fs_basin_activations <- fs_basin %>%
+  left_join(
+    fs_basin_thresholds,
+    by = "HYBAS_ID"
+  ) %>%
+  mutate(
+    fs_above_threshold = sfed >= sfed_threshold
+  ) %>%
+  group_by(
+    basin_id = HYBAS_ID
+  ) %>%
+  mutate(
+    fs_above_threshold_7_days = zoo::rollsum(
+      x = fs_above_threshold,
+      k = 7,
+      align = "left",
+      fill = FALSE
+    ) > 0
+  ) %>%
+  ungroup()
+
+# now do the validations across various metrics
+
+basin_pcts <- basin_activation %>%
+  group_by(
+    basin_id,
+    time
+  ) %>%
+  summarize(
+    across(
+      .cols = `google_activation_2y_rp`:`google_activation_20y_rp`,
+      .fns = mean
+    ),
+    .groups = "drop"
+  ) %>%
+  rename_with(
+    .cols = starts_with("google"),
+    .fn = \(x) str_remove(x, "google_activation_")
+  )
+
+pcts <- seq(.10, .90, by = .1)
+rps <- c(2, 5, 20)
+
+rep_grid <- expand_grid(pcts, rps)
+
+df_basin_metrics <- map2(
+  .x = rep_grid$pcts,
+  .y = rep_grid$rps,
+  .f = \(pct, rp) {
+    basin_valid <- basin_pcts %>%
+      mutate(
+        activation = .data[[paste0(rp, "y_rp")]] >= !!pct
+      ) %>%
+      inner_join(
+        select(
+          fs_basin_activations,
+          basin_id,
+          time,
+          fs_above_threshold_7_days
+        ),
+        by = c("basin_id", "time")
+      ) %>%
+      group_by(
+        basin_id
+      ) %>%
+      mutate(
+        basin_group = cumsum(activation != lag(activation, default = FALSE)),
+        fs_group = cumsum(fs_above_threshold_7_days != lag(fs_above_threshold_7_days, default = FALSE))
+      )
+    
+    basin_positives <- basin_valid %>%
+      filter(
+        activation
+      ) %>%
+      group_by(
+        basin_id,
+        basin_group
+      ) %>%
+      summarize(
+        positive = sum(fs_above_threshold_7_days) > 0,
+        .groups = "drop_last"
+      ) %>%
+      summarize(
+        tp = sum(positive),
+        fp = sum(!positive),
+        .groups = "drop"
+      )
```

Comment:
I think there are a fair amount of cases where the actual % of gauges (`basin_pct`) might fluctuate around the threshold. These cases could potential affect the performance calculations and finding the optimal threshold? I remember this being a tricky aspect of the performance calculations for  the Yemen trigger. I guess the trigger is designed so that the first time it reaches x % the trigger should be activated - do you think this is worth taking into consideration? 

Below is one of these cases if we say the trigger is 80 % (as an example)

```
basin_pcts %>% 
  filter(basin_id==1040022420,
         time %in% seq(as_date("2015-09-16"),
                       as_date("2015-09-16")+4, "day")
  )
```

Reply:
On this, I was thinking that overall, if it goes above and then drops below, we would want to see if there are corresponding floods within 7 days of that alert being generated. And if it rises above again, we would have an alert generated again, so considered them separate, because there is no funding or system. In this way, I wanted to try and basically capture the timing of Google's nowcast data as well.

But I see your point, which is that in our actual analysis, we are looking at the nowcast + 7 day forecasts on each day. And we probably wouldn't alert again in like 3 days (at least with an email, so would be a technicality).

Reply:
So I added in an analysis using the consecutive method, and we generally see the same results. Basically the just barely better metric originally was 90% , and here was 70%. Although really hard to tell if any of this is just random chance or not (I would guess with in-depth analysis we wouldn't be too confident). But given frequency of activation, I'd be reluctant to reduce it from 80% to a lower threshold, particularly as we have only looked at the nowcast and we are going to be triggering off of the nowcast + 7 days in the future (I think we are about to activate according to your latest email!).

Reply:
Overall, really good point and will consider next time for sure!